### PR TITLE
ATS Search UI & API

### PR DESCRIPTION
**Pull request by sanjaytkbabu**
# Description
[PADS-303](https://apps.nrs.gov.bc.ca/int/jira/browse/PADS-303)
[PADS-75](https://apps.nrs.gov.bc.ca/int/jira/browse/PADS-75)
Env variables added to dev, test & prod
Custom environment variables added
New route, controller & service layer for ats
New migration to add ats resource, role, policy to yars
ATSUserSearchParameters backend type added
Submission validator updated
SubmissionForm updated 
New ATSUserLinkModal
New ATSUserDetailModal
New frontend type ATSUser

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/src/validators/enquiry.ts [hunk 1]
```diff
@@ -33,6 +33,7 @@ const schema = {
     body: Joi.object({
       enquiryId: Joi.string().required(),
       activityId: Joi.string().required(),
+      atsClientNumber: Joi.string().allow(null).max(255),
```
Comment (jujaga) [inline app/src/validators/enquiry.ts:36]: SQL Text types do not have an upper bound having a max of 255 isn't really needed unless there's another kind of constraint in the business rules.

File: app/src/db/migrations/20241002000000_011-ats-client-number.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable max-len */
+import type { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+  return (
+    Promise.resolve()
+      // add ats ats_client_number to enquiry table
+      .then(() =>
+        knex.schema.alterTable('enquiry', (table) => {
+          table.text('ats_client_number');
```
Comment (jujaga) [inline app/src/db/migrations/20241002000000_011-ats-client-number.ts:10]: I can't quite remember, but is it nullable by default if not specified? If it is then ignore this comment.

File: frontend/src/components/housing/enquiry/EnquiryForm.vue [hunk 3]
```diff
@@ -278,6 +279,29 @@ onMounted(async () => {
...
+        <InputText
+          v-if="formRef?.values.atsClientNumber"
+          class="col-3 flex-auto"
+          name="atsClientNumber"
+          label="ATS Client #"
+          :disabled="true"
+        />
+        <Button
+          aria-label="Link to ATS"
+          class="h-2rem ml-2"
```
Comment (jujaga) [inline frontend/src/components/housing/enquiry/EnquiryForm.vue:294]: Is this a potential class typo?

File: frontend/src/components/housing/submission/SubmissionForm.vue [hunk 4]
```diff
@@ -609,28 +612,30 @@ onMounted(async () => {
...
+          class="col-3 flex-auto"
           name="atsClientNumber"
           label="ATS Client #"
-          :disabled="!editable"
+          :disabled="true"
         />
-        <div class="col-9" />
+        <Button
+          aria-label="Link to ATS"
+          class="h-2rem ml-2"
```
Comment (jujaga) [inline frontend/src/components/housing/submission/SubmissionForm.vue:626]: Is this a potential class typo?

File: frontend/src/services/atsService.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { appAxios } from './interceptors';
+
+export default {
+  searchATSUsers(params?: any) {
```
Comment (jujaga) [inline frontend/src/services/atsService.ts:4]: Missing JSDocs

File: app/config/custom-environment-variables.json [hunk 6]
```diff
@@ -30,6 +30,13 @@
   },
   "server": {
     "apiPath": "SERVER_APIPATH",
+    "ats": {
+      "apiPath": "SERVER_ATS_APIPATH",
+      "tokenUrl": "SERVER_ATS_TOKENURL",
+      "clientId": "SERVER_ATS_CLIENTID",
+      "clientSecret": "SERVER_ATS_CLIENTSECRET",
+      "integration": "SERVER_ATS_INTEGRATION"
```
Comment (kyle1morel) [inline app/config/custom-environment-variables.json:38]: Please alphabetize

File: app/src/db/migrations/20241002000000_011-ats-integration.ts [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+  },
+  {
+    name: Action.DELETE
+  }
+];
+
+export async function up(knex: Knex): Promise<void> {
+  return (
+    Promise.resolve()
+      // add ats ats_client_number to enquiry table
```
Comment (kyle1morel) [inline app/src/db/migrations/20241002000000_011-ats-integration.ts:30]: Please capitalize comments for consistency with the rest of the base

File: app/src/services/ats.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+//       client_id: config.get('server.ats.clientId'),
+//       client_secret: config.get('server.ats.clientSecret')
+//     },
+//     headers: {
+//       'Content-type': 'application/x-www-form-urlencoded'
+//     },
+//     withCredentials: true
+//   });
+//   return response.data.access_token;
+// }
```
Comment (kyle1morel) [inline app/src/services/ats.ts:28]: Please remove commented code if it is no longer required.

File: app/src/validators/enquiry.ts (hunk 1 above)
Comment (kyle1morel) [reply app/src/validators/enquiry.ts:36]: This is consistent with the rest of the base. Limit via the API as we're using `text` in the db.

File: frontend/src/components/housing/enquiry/EnquiryForm.vue [hunk 9]
```diff
@@ -278,6 +279,29 @@ onMounted(async () => {
...
+          @click="
+            () => {
+              atsUserLinkModalVisible = true;
+            }
+          "
```
Comment (kyle1morel) [inline frontend/src/components/housing/enquiry/EnquiryForm.vue:299]: An inline function is not required. Can just do `@click="atsUserLinkModalVisible = true"`

File: frontend/src/components/housing/enquiry/EnquiryForm.vue [hunk 10]
```diff
@@ -327,4 +351,13 @@ onMounted(async () => {
       </div>
     </div>
   </Form>
+  <ATSUserLinkModal
+    v-model:visible="atsUserLinkModalVisible"
+    @ats-user-link:link="
+      (atsUser: ATSUser) => {
+        formRef?.setFieldValue('atsClientNumber', atsUser.atsClientNumber);
```
Comment (kyle1morel) [inline frontend/src/components/housing/enquiry/EnquiryForm.vue:358]: Please add `v-slot="{ setFieldValue }"` to the `<Form>` and remove `formRef?.`. The formRef should only be used inside of `<script>`.

File: frontend/src/components/housing/submission/SubmissionForm.vue [hunk 11]
```diff
@@ -609,28 +606,30 @@ onMounted(async () => {
...
-
-      <div
-        v-if="values.addedToATS"
-        class="w-full"
-      >
+      <SectionHeader title="ATS" />
+      <div class="flex align-items-center mb-2">
         <InputText
-          class="col-3"
+          v-if="formRef?.values.atsClientNumber"
```
Comment (kyle1morel) [inline frontend/src/components/housing/submission/SubmissionForm.vue:612]: Please use the `values` from the `v-slot` over the `formRef`.

File: frontend/src/components/housing/submission/SubmissionForm.vue [hunk 12]
```diff
@@ -705,4 +704,14 @@ onMounted(async () => {
       </div>
     </div>
   </Form>
+  <ATSUserLinkModal
+    v-model:visible="atsUserLinkModalVisible"
+    @ats-user-link:link="
+      (atsUser: ATSUser) => {
+        console.log(atsUser);
+        formRef?.setFieldValue('atsClientNumber', atsUser.atsClientNumber);
```
Comment (kyle1morel) [inline frontend/src/components/housing/submission/SubmissionForm.vue:712]: Please see my other comment regarding `setFieldValue`.

File: frontend/src/components/user/ATSUserLinkModal.vue [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+// State
+const loading: Ref<boolean> = ref(false);
+const firstName: Ref<string> = ref('');
+const lastName: Ref<string> = ref('');
+const atsClientNumber: Ref<string> = ref('');
+const phone: Ref<string> = ref('');
+const email: Ref<string> = ref('');
+const selectedUser: Ref<ATSUser | undefined> = ref(undefined);
+const users: Ref<Array<ATSUser>> = ref([]);
+const visible = defineModel<boolean>('visible');
```
Comment (kyle1morel) [inline frontend/src/components/user/ATSUserLinkModal.vue:24]: Alphabetize

File: frontend/src/components/user/ATSUserLinkModal.vue [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    :draggable="false"
+    :modal="true"
+    class="app-info-dialog w-8"
+  >
+    <template #header>
+      <span class="p-dialog-title">Search ATS</span>
+    </template>
+
+    <div class="pt-1 mb-4 flex justify-content-between">
+      <div class="col-11.5 align-content-center">
```
Comment (kyle1morel) [inline frontend/src/components/user/ATSUserLinkModal.vue:73]: I don't believe half `col`s exist. They must be whole numbers. At least I don't see them in the https://primeflex.org/gridsystem docs

File: frontend/src/types/ATSUser.ts [hunk 15]
```diff
@@ -0,0 +1,8 @@
+export type ATSUser = {
+  email: string;
+  firstName: string;
+  surName: string;
+  atsClientNumber: string | undefined;
+  phone: string;
+  address: string;
```
Comment (kyle1morel) [inline frontend/src/types/ATSUser.ts:7]: Alphabetize

**Comment by kyle1morel** [comment]
Please include the ticket number in the description.

**Comment by sanjaytkbabu** [comment]
> Please include the ticket number in the description.

done!

File: frontend/src/components/housing/enquiry/EnquiryForm.vue (hunk 3 above)
Comment (sanjaytkbabu) [reply frontend/src/components/housing/enquiry/EnquiryForm.vue:295]: it exists!: https://primeflex.org/height

File: app/src/services/ats.ts (hunk 8 above)
Comment (sanjaytkbabu) [reply app/src/services/ats.ts]: done

**Pull request merged**